### ISSUE TITLE
[BUG] - 1.35.* v7 nodes on Mainnet are picking up the wrong parent-block - Vasil HF Blocker?

### ISSUE THREAD

**Issue by gitmachtl** [input-output-hk/cardano-node#4228]
### Strange Issue

So, i will post a bunch of pics here, screenshots taken from pooltool.io. Blockflow is from botton->top. These are only a few examples, there are many more!

**This is happening on mainnet right now** running 1.35.* nodes. There are many occasions with double & tripple height battles where **newer v7 1.35.\* nodes** are **picking up the wrong parent-block** and try to build on another v7 block. So v6 1.34.* nodes are winning those all the time. 

I personally had the issue loosing 10 height-battles within a 1-2 day window against v6 nodes. That was 100% of all my height-battles i had. 

Its always the same pattern: There is a height battle that a v7 node looses against a v6 node. If there are also two nodes scheduled for the next block and one of them is a v7 node, it picks up the wrong lost block hash from the previous v7 node and builds on it. Of course it looses against the other v6 node which is building on the correct block. But as you can see in the example below, this can span multiple slotheights/blocks ⚠️ 

This is a Vasil-HF blocker IMO, because it would lead to the situation that SPOs are only upgrading to 1.35.* at the last possible moment before the HF, giving the ones staying on 1.34.1 an advantage. Not a good idea, it must be sorted out before. Q&A team please start an investigation on that asap, thx! 🙏 

![image](https://user-images.githubusercontent.com/47434720/180648823-942ce127-18d8-4aa7-b5c5-b2506852482f.png)
-----
![image](https://user-images.githubusercontent.com/47434720/180648826-f0317024-d6d1-48f7-8d28-03a7cce45c14.png)
-----
Here is a nightmare one, v7 built ontop of other v7 node (green path):
![image](https://user-images.githubusercontent.com/47434720/180649301-95f087c4-c882-4909-bb56-6ecc33447a15.png)
-----
![image](https://user-images.githubusercontent.com/47434720/180648837-4063c8c9-3746-4e49-9cb9-9b0796d21722.png)
-----
![image](https://user-images.githubusercontent.com/47434720/180648850-e9892b2d-1538-45c4-a1cd-fbcd18bb764f.png)
-----
![image](https://user-images.githubusercontent.com/47434720/180649488-3c6cdc63-4e9e-4dcc-a846-461d6af7d581.png)
-----
![image](https://user-images.githubusercontent.com/47434720/180648859-8a0b757e-d5b5-4b8a-8e46-ea7623b2b714.png)
-----
![image](https://user-images.githubusercontent.com/47434720/180648865-5589296a-9ba8-47fa-9635-603ad5264a01.png)

**Comment by papacarp** [comment]
height battles are on the rise.  I'll work on analyzing chained height battles also but wanted to drop this data here as part of it.

![image](https://user-images.githubusercontent.com/11194546/180657790-3c5747cf-5e7b-41ac-aa14-bde97a0b4cdd.png)

**Comment by njd42** [comment]
@papacarp I note that pooltool.io/realtime is giving surprisingly number of nodes for certain blocks (i.e less than ten) - this seems unusual. 

Does anyone have evidence (i.e log entries) that show that any of these "height battles" above occurred _when_ the superseded block had been received by the about-to-forge node? My working hypothesis would be some sort of connectivity partition, I'm looking for evidence that assumption is incorrect.

**Comment by njd42** [comment]
@gitmachtl - can I ask, are all these 'V7' systems deployed using the same scripts? Just seeing if there could be a common configuration issue that underlies this.

**Comment by gitmachtl** [comment]
@njd42 as far as i can tell and i have spoken to a bunch of those SPOs, they are all running completely different. different deploy scripts, different locations, different kind of hardware (vm, dedicated server, ...). 

the number of nodes reporting are only the ones that are actively reporting them to pooltool, but the height battles can be confirmed by others too.

**Comment by rene84** [comment]
For what it's worth: we are running our own compiled node in Docker (own Dockerfile) on AWS ECS and migrated to 1.35.1 (too) early.

Haven't done the amount of investigation that @gitmachtl has so I can only share sparse findings now, but what I could see is that our BP would mint a block omitting the preceding block (thus getting dropped by the next leader) even though there was sometimes 10+ slots in between. Our relays never received that preceding block (checked logs) so naturally the BPs also didn't. Will keep an eye on this issue and report when I've done more checks

**Comment by renesecur** [comment]
Example: block 7530726

Block 7530725 arrived at our BP:

```
{
    "app": [],
    "at": "2022-07-22T12:25:00.17Z",
    "data": {
        "chainLengthDelta": 1,
        "kind": "TraceAddBlockEvent.AddedToCurrentChain",
        "newtip": "4e4e385bae46bc9edb06c5203a70ccf64dabc64724aad4f4256f1f99cbaaf80d@66926409"
    },
    "env": "1.35.1:c7545",
    "host": "ip-172-2",
    "loc": null,
    "msg": "",
    "ns": [
        "cardano.node.ChainDB"
    ],
    "pid": "82",
    "sev": "Notice",
    "thread": "12135"
}
```
Then our block minted and reached our relays (three each on 1.35.1, on with p2p enabled). Then 18 slots later the block from 4f3410f074e7363091a1cc1c21b128ae423d1cce897bd19478e534bb arrived with hash 3150a0d4502f1d0995c4feae3fcb93dd505671cf8dc84bbf0b761e2ee64d70dc and our relays switched fork.

<img width="1171" alt="image" src="https://user-images.githubusercontent.com/89156350/180662747-77e0d51e-3bb7-43ed-ba58-e07c34ef8cbe.png">

The number of incoming connections on our relay is healthy and our block propagation seems otherwise fine.

**Comment by andreabravetti** [comment]
@rene84 what if you have two relay, one in 1.34.x and one in 1.35.x? does your bp receive that preceding block?

**Comment by njd42** [comment]
I think that the connectivity partition is a much more likely root cause than an issue with chain selection given the evidence at hand. However, such a partition is _highly_ unlikely to occur through pure random chance, which suggests some sort of trigger. 

If you scan through your logs looking for, say `ErrorPolicySuspendPeer` messages? What we are looking for is a rash of those occurring at around the same time causing the 1.34.x and 1.35.x nodes to drop connections with each other.

**Comment by coot** [comment]
Has this behaviour been observed on non P2P 1.35.2 nodes, or is this limited to P2P only nodes?

**Comment by nemo83** [comment]
Great work @gitmachtl I'm wondering if the issue you're reporting is somehow related to https://github.com/input-output-hk/cardano-node/issues/4226.

In my case, given the importance of the upgrade, I only bumped up the version of one of my co-located relays. What I've noticed is that, multiple times, the BP running 1.34.1 would _stop talking_ to the relay running 1.35.1. Only way to restore was to bounce the relay on 1.35.1.

It looks like 1.35.1 nodes eventually _stop_ talking to 1.34.1 and maybe start generating heights battles?

**Comment by gitmachtl** [comment]
> Has this behaviour been observed on non P2P 1.35.2 nodes, or is this limited to P2P only nodes?

As far as i know the SPOs were running 1.35.1 nodes in non p2p mode on mainnet. It is really a strange thing that 1.35.* nodes keep on sticking with the block of other 1.35.* nodes and not picking up a resolved 1.34.* block. 1.35.2 was not used i guess, its just too fresh out of the box.

**Comment by gitmachtl** [comment]
> Great work @gitmachtl I'm wondering if the issue you're reporting is somehow related to #4226.
> 
> In my case, given the importance of the upgrade, I only bumped up the version of one of my co-located relays. What I've noticed is that, multiple times, the BP running 1.34.1 would _stop talking_ to the relay running 1.35.1. Only way to restore was to bounce the relay on 1.35.1.
> 
> It looks like 1.35.1 nodes eventually _stop_ talking to 1.34.1 and maybe start generating heights battles?

Could be, but the same relays/BPs later on are producing normal blocks.

**Comment by renesecur** [comment]
These are the occurrences of `ErrorPolicySuspendPeer` on the three relays on the 22nd of July. First to were on 1.35.1 without p2p, the third one was on p2p

<img width="1049" alt="image" src="https://user-images.githubusercontent.com/89156350/180724369-12cf818f-4c80-4384-aed8-64c6ee939647.png">
<img width="1050" alt="image" src="https://user-images.githubusercontent.com/89156350/180724418-38424da2-f55a-4c53-a719-e756376c9e13.png">
<img width="1064" alt="image" src="https://user-images.githubusercontent.com/89156350/180724468-87bcf6a1-321e-4093-aa73-02a3f05abc39.png">

In UTC

**Comment by gitmachtl** [comment]
I am not sure if we're mixing up two different issues here. I would love to hear @coot 's oppinion on the height battle issues on the mainnet, its really outstanding and not just a glitch.

**Comment by coot** [comment]
We haven't seen such long heights battles before, it certainly deserves attention.

**Comment by gitmachtl** [comment]
The strange thing is, that with those double height battles, the 1.35.* only builds on the lost block of the 1.35.* before. Normally it should pick up the block from the previous winner of course. And the fact that it happend than again is really bad. Creating mini forks all the time. I was on a full 1.35.1 setup at this time on mainnet, relays and BPs. I reverted back to a full 1.34.1 setup and had no issues like that anymore since than. I know that this is hard to simulate, because on the Testnet we now are in babbage-era and all nodes are 1.35.*. On the Mainnet we have the mixed situation. But SPOs are now aware, and it could lead to the case that SPOs are only updating in the very last moment to stay out of such bugs. Thats not a good solution to go into this HF i would say.

**Comment by coot** [comment]
> These are the occurrences of `ErrorPolicySuspendPeer` on the three relays on the 22nd of July. First to were on 1.35.1 without p2p, the third one was on p2p
> 

`P2P` nodes use a different trace message, you'd need to search for `TraceDemoteAsynchronous` and `Error` as [here](https://github.com/input-output-hk/cardano-node/blob/master/cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs?plain=1#L1773-L1779=) (we ought to change it to something better, `Error` is terrible for grepping).

**Comment by rene84** [comment]
What is consistent with what I see in the logs is that some blocks just never make it even after 10 or more seconds. Both incoming and outgoing I've seen that. Seems more likely due to a propagation issue in the network eg 1.34.x and 1.35.x not talking rather then there being a bug in the fork selection mechanism in the node itself

**Comment by coot** [comment]
Can I ask for logs from a block producer which was connected to a `1.35.x` node from the time when it lost a slot battle.  It would help us validate our hypothesis.  If one doesn't want to share them publicly here, you can send them to my iohk email address (`marcin.szamotulski@iohk.io`) or share them through discord.

**Comment by rene84** [comment]
Sure. Afk atm so I can share that in about 5 hours from now. Note that I can only help you with BP on 1.35.1 connected to a relay on 1.35.1

**Comment by coot** [comment]
> Sure. Afk atm so I can share that in about 5 hours from now. Note that I can only help you with BP on 1.35.1 connected to a relay on 1.35.1

I am not sure if this will help, but please send.  It's more interesting to see `1.34.1` BP connected to `1.35.1` relay.

**Comment by gitmachtl** [comment]
here is another example:

![image](https://user-images.githubusercontent.com/47434720/180801254-ddf46c03-1b01-4599-ad4c-6aa5ddd5fac9.png)

**Comment by gitmachtl** [comment]
and another one, in this example i know for sure that PILOT is running 1.35.0 on all mainnet nodes in normal mode (not p2p)

![image](https://user-images.githubusercontent.com/47434720/180802744-16449d6a-dfc5-4246-8780-0798eaee1fed.png)

**Comment by renesecur** [comment]
> > These are the occurrences of `ErrorPolicySuspendPeer` on the three relays on the 22nd of July. First to were on 1.35.1 without p2p, the third one was on p2p
> 
> `P2P` nodes use a different trace message, you'd need to search for `TraceDemoteAsynchronous` and `Error` as [here](https://github.com/input-output-hk/cardano-node/blob/master/cardano-node/src/Cardano/Tracing/OrphanInstances/Network.hs?plain=1#L1773-L1779=) (we ought to change it to something better, `Error` is terrible for grepping).

I searched for `TraceDemoteAsynchronous` but did not get any results. I'm pretty confident we haven't enabled the correct trace for that log message.  I do see regular `PeerSelectionCounters` and the number of hotPeers changing from time to time but not `Error` log message on the entire day of the 22nd of July (UTC)

**Comment by coot** [comment]
@renesecur you need to enable `TracePeerSelection` and `TracePeerSelectionActions`.

**Comment by coot** [comment]
We would like to get logs from a `1.34.1` relay connected to a `1.35.x` BP node which was part of any of the slot battles.

**Comment by coot** [comment]
One more favour to ask, those of you who are running `1.35.x` BP, could you share your `1.35.x` relay? We'd like to connect a `1.34.1` relay and observe what's going on on our end.

**Comment by feqifei** [comment]
> One more favour to ask, those of you who are running `1.35.x` BP, could you share your `1.35.x` relay? We'd like to connect a `1.34.1` relay and observe what's going on on our end.

eu1.stakecool.io:4001 , 1.35.2 p2p enabled
eu2.stakecool.io:4001 , 1.35.2 p2p disabled
eu3.stakecool.io:4002 , 1.34.1 p2p disabled

They are serving six 1.35.2 block producers. Let me know if you want me to add your relay

**Comment by karknu** [comment]
> height battles are on the rise. I'll work on analyzing chained height battles also but wanted to drop this data here as part of it.
> 
> ![image](https://user-images.githubusercontent.com/11194546/180657790-3c5747cf-5e7b-41ac-aa14-bde97a0b4cdd.png)

@papacarp  Great stats. 
Could you make a plot that tracks slotbattles / (total block bytes in epoch) and heightbattles / (toal block bytes in epoch) ? 
The number of heights battles is expected to increase with increased utilization and looking back at earlier epochs they where more common a few months ago.

**Comment by coot** [comment]
> eu1.stakecool.io:4001 , 1.35.2 p2p enabled
> eu2.stakecool.io:4001 , 1.35.2 p2p disabled
> eu3.stakecool.io:4002 , 1.34.1 p2p disabled

@feqifei thanks.  Have you observed any invalid block on the `1.34.1` relay so far?

**Comment by gitmachtl** [comment]
> One more favour to ask, those of you who are running `1.35.x` BP, could you share your `1.35.x` relay? We'd like to connect a `1.34.1` relay and observe what's going on on our end.

I posted it in the SPO best practice group, lets see if we can collect some full 1.35.* setups here.

**Comment by feqifei** [comment]
> > eu1.stakecool.io:4001 , 1.35.2 p2p enabled
> > eu2.stakecool.io:4001 , 1.35.2 p2p disabled
> > eu3.stakecool.io:4002 , 1.34.1 p2p disabled
> 
> @feqifei thanks. Have you observed any invalid block on the `1.34.1` relay so far?

not anymore since 2022-07-23 8:00 utc. My BP lost several blocks due to bad/wrong propagation only during that 24 hrs starting from 2022-07-22 7:30 utc. Unfortunately I have no logs related to that timeframe

**Comment by stefiix92** [comment]
> > One more favour to ask, those of you who are running `1.35.x` BP, could you share your `1.35.x` relay? We'd like to connect a `1.34.1` relay and observe what's going on on our end.
> 
> I posted it in the SPO best practice group, lets see if we can collect some full 1.35.* setups here.

18.202.232.130
3.249.180.234
3.249.41.89
54.74.196.6

all relays are running on 1.35.0 with port 3001. All BP nodes were upgraded to 1.35.0. In the topology, we have 4/7 relays running on 1.35.0, the left 3 are running on 1.34.1

**Comment by TheScenePool** [comment]
> One more favour to ask, those of you who are running `1.35.x` BP, could you share your `1.35.x` relay? We'd like to connect a `1.34.1` relay and observe what's going on on our end.
Relay1.thescenepool.com 6000
Relay2.thescenepool.com 6000
Relay3.thescenepool.com 6000
Relay4.thescenepool.com 6000

All running 1.35.0

Note my BP will not produce blocks due to low delegation but hopefully you can see something.

**Comment by gitmachtl** [comment]
It happend again
![image](https://user-images.githubusercontent.com/47434720/181000535-551f130b-9fc1-43c9-8d86-89ff3a77af26.png)

**Comment by feqifei** [comment]
one more case
![image](https://user-images.githubusercontent.com/42180581/181003693-95b1946e-78cf-4b1b-a57c-7254a831e8de.png)
We notice that not-adopted blocks are propagated by very few nodes (between 50 and 60).
We suspect that those 50-60 nodes are those ones that already moved to 1.35.x version, but we have no proof.

Maybe @papacarp can give us more information about those 53 nodes that report block hash 5ab7d7741fce303623790f551be495375728a61bb9febf1a2e81dcb1948b2091

**Comment by feqifei** [comment]
![image](https://user-images.githubusercontent.com/42180581/181005315-59aca6a5-d0f4-4036-ab0e-7c6ef8559bde.png)

**Comment by feqifei** [comment]
![image](https://user-images.githubusercontent.com/42180581/181006139-900298a9-f1cc-467d-a4c2-7ad01dcb59c1.png)

**Comment by stefiix92** [comment]
> Maybe @papacarp can give us more information about those 53 nodes that report block hash 5ab7d7741fce303623790f551be495375728a61bb9febf1a2e81dcb1948b2091

my BP node's (1.35.0) log for 5ab7d7741fce303623790f551be495375728a61bb9febf1a2e81dcb1948b2091 block:
```
[4eefd458:cardano.node.ChainDB:Notice:220] [2022-07-26 12:13:55.14 UTC] Chain extended, new tip: 5ab7d7741fce303623790f551be495375728a61bb9febf1a2e81dcb1948b2091 at slot 67271343
```

Then why others blocks (i.e. 9be22a26e9a650813b37ef457ecec6b7b8ac2493d865f9f5b5ec63ffb0b5f0d7) were minted on v7 and confirmed by 290 nodes? Or does it involve only Height battles?

**Comment by karknu** [comment]
If anyone is running a relay with a pre 1.35.x version it would be interesting to see how that node deals with a V7 block.

`68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9` is a suspected v7 block that lost out to `80572fdeb5dd826c56ff4cae57ce2d0b17b0fd6aed1cfac692431a558293fe5e`.  
Grep for `68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9`  and look for errors. If you are running with TraceChainSyncClient enabled the `TraceDownloadedHeader` will give you the addresses of the peers that presented your node with that block. Grep for that ip addresses to see what your node does with that peer.

**Comment by JaredCorduan** [comment]
if anyone is able to give me the cbor bytes (hex would be nice) of one of these v7 blocks that loses such a height battle, I can check to see if there is any incompatibility in the serialization between v6 and v7.

**Comment by feqifei** [comment]
> If anyone is running a relay with a pre 1.35.x version it would be interesting to see how that node deals with a V7 block.
> 
> `68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9` is a suspected v7 block that lost out to `80572fdeb5dd826c56ff4cae57ce2d0b17b0fd6aed1cfac692431a558293fe5e`. Grep for `68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9` and look for errors. If you are running with TraceChainSyncClient enabled the `TraceDownloadedHeader` will give you the addresses of the peers that presented your node with that block. Grep for that ip addresses to see what your node does with that peer.

I see the 68c39 block in my two 1.35.2 relays but not in my two 1.34.1 relays

**Comment by karknu** [comment]
> > If anyone is running a relay with a pre 1.35.x version it would be interesting to see how that node deals with a V7 block.
> > `68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9` is a suspected v7 block that lost out to `80572fdeb5dd826c56ff4cae57ce2d0b17b0fd6aed1cfac692431a558293fe5e`. Grep for `68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9` and look for errors. If you are running with TraceChainSyncClient enabled the `TraceDownloadedHeader` will give you the addresses of the peers that presented your node with that block. Grep for that ip addresses to see what your node does with that peer.
> 
> I see the 68c39 block in my two 1.35.2 relay logs but not in my two 1.34.1 relay logs

Interesting. Do you have chaindb and chainsync client tracers enabled on the 1.34.1 relays? 
```
  "TraceChainDb": true,
  "TraceChainSyncClient": true,
```

**Comment by feqifei** [comment]
> 

ChainDB true, ChainSyncClient false.
both 1.34.1 nodes were busy on checking some DNS subscriptions, they didn't receive any block. 42 seconds later both received the adversary block 80572...

**Comment by cardanoinvest** [comment]
95.183.53.20:6060
95.217.222.194:6060
My 1.35.0 relays, for my 1.35.0 BP.  I have 1 block on 2022-07-27T02:55:02+02:00 .
Do you want me to run some extra logs? I made 3 blocks with the same setup during previous epoch, no issues.

**Comment by gitmachtl** [comment]
> one more case, MS9 on v7, LEAF on v6 ![image](https://user-images.githubusercontent.com/42180581/181003693-95b1946e-78cf-4b1b-a57c-7254a831e8de.png) We noticed that not-adopted blocks are propagated by very few nodes (between 50 and 60). We suspect that those 50-60 nodes are those ones that already moved to 1.35.x version, but we have no proof.

yeah thats because its the number of nodes reporting to pooltool running 1.35.*, so this signals that only the other v7 nodes are getting/reporting the block

**Comment by feqifei** [comment]
@karknu 
About your example I think that the reason why 1.34.1 nodes didn't receive information about block 68c39e855361e637ee76862d59dc04eb190398ae33cae7cbcd1b6f6cd921fac9 is because the node already marked as invalid its predecessor 3677ae1bb3697402f443a737a18eab5194c66f2efaec1ff2c2bc0122413d9437, forged by another v7 BP.

![image](https://user-images.githubusercontent.com/42180581/181047420-65f0c414-4e76-4f72-aeed-ca9234aedf08.png)

So, I suppose the 1.34.1 already discharged that fork and didn't consider it.
More interesting is the error why 1.34.1 decided to consider invalid 3677...

{"host":"ff01-rl0","pid":"448636","loc":null,"at":"2022-07-26T12:25:28.73Z","ns":["cardano.node.ChainDB"],"sev":"Error","env":"1.34.1:73f9a","data":{"kind":"TraceAddBlockEvent.AddBlockValidation.InvalidBlock","error":"ExtValidationErrorLedger (HardForkLedgerErrorFromEra S (S (S (S (Z (WrapLedgerErr {unwrapLedgerErr = BBodyError (BlockTransitionError [ShelleyInAlonzoPredFail (LedgersFailure (LedgerFailure (UtxowFailure (WrappedShelleyEraFailure (UtxoFailure (FeeTooSmallUTxO (Coin 1006053) (Coin 1001829))))))),ShelleyInAlonzoPredFail (LedgersFailure (LedgerFailure (UtxowFailure (WrappedShelleyEraFailure (UtxoFailure (FeeTooSmallUTxO (Coin 1011641) (Coin 1007417)))))))])}))))))","block":{"hash":"3677ae1bb3697402f443a737a18eab5194c66f2efaec1ff2c2bc0122413d9437","kind":"Point","slot":67272037}},"msg":"","thread":"406","app":[]}

Instead, 1.35.2 node considered this block (or at least its header) as valid 
{"app":[],"at":"2022-07-26T12:25:28.24Z","data":{"chainLengthDelta":1,"kind":"TraceAddBlockEvent.AddedToCurrentChain","newtip":"3677ae1bb3697402f443a737a18eab5194c66f2efaec1ff2c2bc0122413d9437@67272037"},"env":"1.35.2:7612a","host":"gv01-rl0","loc":null,"msg":"","ns":["cardano.node.ChainDB"],"pid":"2554846","sev":"Notice","thread":"400"}

**Comment by JaredCorduan** [comment]
@feqifei do you have a way of getting the serialized block corresponding to `3677ae1bb3697402f443a737a18eab5194c66f2efaec1ff2c2bc0122413d9437` ? that would allow me to investigate if there is a difference in the fee calculation between 1.34 and 1.35.

**Comment by feqifei** [comment]
> @feqifei do you have a way of getting the serialized block corresponding to `3677ae1bb3697402f443a737a18eab5194c66f2efaec1ff2c2bc0122413d9437` ? that would allow me to investigate if there is a difference in the fee calculation between 1.34 and 1.35.

Hi Jared, I don't know how to satisfy your request but consider that this is a block made by another BP (not mine) and it didn't reach the blockchain. So, even if I had enough skill to answer your question (and I have not :) ) I really don't know how to collect this info

**Comment by JaredCorduan** [comment]
ok, no problem @feqifei , your report above is already very helpful, it gives me a new lead!

**Comment by papacarp** [comment]
I have now correlated the reporter versions along with the battles. This is not on the public website yet, but I captured one during development that I think tells us what we need to know.  The reporters of the orphaned blocks are generally on 1.35.  Note that eventually all nodes will report the winning block because they have to.  But if you watch this realtime you'll see initially no 1.34 nodes report the 1.35 block.

Next step is to get the block protocol versions into this now that cncli was upgraded to report that info as well.  so we will know if the orphan block (as well as the chained block) are produced by a 1.35 node

![image](https://user-images.githubusercontent.com/11194546/181053254-a000ebce-ba22-4067-8e04-4e0170f006f0.png)

**Comment by gitmachtl** [comment]
FeeTooSmall issue!? Hmm... so maybe a complex thing with smart contracts involved? And those v7 blocks that are getting accepted normally by v6 nodes don't have smart contracts in them or luckily enough fees set?

**Comment by JaredCorduan** [comment]
> FeeTooSmall issue!? Hmm... so maybe a complex thing with smart contracts involved? And those v7 blocks that are getting accepted normally by v6 nodes don't have smart contracts in them or luckily enough fees set?

I strongly suspect the bug is something like that :point_up: . I'm investigating this now, just comparing the code. A reproducible would be the most helpful thing right now, ie a serialized v7 block that lost one of these height battles.

**Comment by gitmachtl** [comment]
yep, but hard to capture and store for a "normal spo" i guess.

**Comment by gitmachtl** [comment]
@JaredCorduan could it the utxoCostPerByte conversion? Because its now 4310 lovelaces per byte thats 34480 lovelaces per word. But the current parameters on the mainnet are set to 34482 lovelaces.

**Comment by JaredCorduan** [comment]
@gitmachtl I don't think so, that only kicks in with the Babbage era, and we are seeing this on mainnet (alonzo).

**Comment by gitmachtl** [comment]
The calculation is based on the bytes in alonzo, but maybe 1.35.0/1.35.1 nodes are using it internally with the *8 to get to the utxoCostPerWord value also for the alonzo era?

**Comment by Eztero** [comment]
> One more favour to ask, those of you who are running `1.35.x` BP, could you share your `1.35.x` relay? We'd like to connect a `1.34.1` relay and observe what's going on on our end.

I have the producer and the relay in the version 1.35.0

relay:
51.222.156.238 : 3001

**Comment by karknu** [comment]
@JaredCorduan https://gist.github.com/karknu/3ba79778b83f35d43789ff68436b114f is a cbor of the loosing block for the height battle for block number 7547634 . It is manually stitched together from packets in a pcap so may not be 100% correct.

**Comment by gitmachtl** [comment]
From Conrad:

relay01.bladepool.com:3001
relay02.bladepool.com:3001 
Current set up:
BP: 1.35.2
Relays: 1.34.1 and 1.35.2 (mixed environment)

**Comment by gitmachtl** [comment]
> @gitmachtl I don't think so, that only kicks in with the Babbage era, and we are seeing this on mainnet (alonzo).

Maybe some strange utf encoding/decoding issues for metadata and co?

**Comment by papacarp** [comment]
> @papacarp I note that pooltool.io/realtime is giving surprisingly number of nodes for certain blocks (i.e less than ten) - this seems unusual.
> 
> Does anyone have evidence (i.e log entries) that show that any of these "height battles" above occurred _when_ the superseded block had been received by the about-to-forge node? My working hypothesis would be some sort of connectivity partition, I'm looking for evidence that assumption is incorrect.

I watched a block get 50 reports and then like 10 seconds later a new block came in and it orphaned the first block.  Seems like the 1.34 nodes never got the 1.35 block so we only got reports from the 1.35 nodes.  Then once a new block came in the 1.34 nodes moved the chain forward.

I think a network partition is the right angle to explore.  1.34 nodes stop talking to 1.35 nodes.

**Comment by gitmachtl** [comment]
So the scenario is: 1.35.* node is producing a block, for some reason a 1.34.* (or below) node sees this block as an invalid block and rejects the block and in addition terminates the connection to the 1.35.* node. Only other 1.35.* nodes are seeing the made block as valid and continue to build on them.

**Comment by nemo83** [comment]
> > @papacarp I note that pooltool.io/realtime is giving surprisingly number of nodes for certain blocks (i.e less than ten) - this seems unusual.
> > Does anyone have evidence (i.e log entries) that show that any of these "height battles" above occurred _when_ the superseded block had been received by the about-to-forge node? My working hypothesis would be some sort of connectivity partition, I'm looking for evidence that assumption is incorrect.
> 
> I watched a block get 50 reports and then like 10 seconds later a new block came in and it orphaned the first block. Seems like the 1.34 nodes never got the 1.35 block so we only got reports from the 1.35 nodes. Then once a new block came in the 1.34 nodes moved the chain forward.
> 
> I think a network partition is the right angle to explore. 1.34 nodes stop talking to 1.35 nodes.

Hi, I've opened a bug exzctly for this reason. https://github.com/input-output-hk/cardano-node/issues/4226

My BP is 1.34.1 and I've got 1 relay on 1.35.1. I've seen the BP periodically loosing connection to 1.35.1 and failing to restore it. You can check the logs provided in the ticket for more info. I really hope thi helps! 

Great work everyone!

I did update to 1.35.2 and also in this case I have experienced the BP to drop connection to 1.35.2 and never re-enstate it (unless I bump the relay.)

**Comment by JaredCorduan** [comment]
I think I have found the problem. If I'm correct, it is a difference between how 1.34 and 1.35 compute the minimum fee.

There are two functions named `minfee` with nearly the same signature:

https://github.com/input-output-hk/cardano-ledger/blob/14e1bcc89e275600efb8b66c7cefeebfb1764204/eras/shelley/impl/src/Cardano/Ledger/Shelley/Tx.hs#L586-L597

https://github.com/input-output-hk/cardano-ledger/blob/14e1bcc89e275600efb8b66c7cefeebfb1764204/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs#L322-L340

We refactored our rules post node 1.34, attempting to reduce code copy, but ended up placing the Shelley calculation in the Alonzo rules:

https://github.com/input-output-hk/cardano-ledger/blob/14e1bcc89e275600efb8b66c7cefeebfb1764204/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs#L288

The difference between the two calculations is that the Alonzo one takes the script execution into account. This means that 1.34 nodes accept a (strict) subset of blocks that 1.35 deems valid. So what we have is a soft fork where the "conservative nodes" (ie 1.34) are winning the longest chain.

---

In hindsight, we should have had golden tests for the fee calculation in each era (we have them for some eras, but it seems we missed the Alonzo ones). I do not understand yet why our property tests did not catch this, as it seems to be using the correct `minfee`:

https://github.com/input-output-hk/cardano-ledger/blob/14e1bcc89e275600efb8b66c7cefeebfb1764204/eras/alonzo/test-suite/src/Test/Cardano/Ledger/Alonzo/AlonzoEraGen.hs#L34-L39

**Comment by JaredCorduan** [comment]
If I am able to push out a quick fix on top of node 1.35.2, would y'all feel confident in your ability to see if the problem if fixed by my patch on the testnet? In the meantime I will think about how best to test this properly.

**Comment by gitmachtl** [comment]
Thx @JaredCorduan , this sounds reasonable. Glad that i am not crazy as some said to me ... 😆

**Comment by nemo83** [comment]
> ~If I am able to push out a quick fix on top of node 1.35.2, would y'all feel confident in your ability to see if the problem is fixed by my patch on the testnet? In the meantime I will think about how best to test this properly.~
> 
> Never mind, Babbage is actually using the correct `minfee` calculation: https://github.com/input-output-hk/cardano-ledger/blob/14e1bcc89e275600efb8b66c7cefeebfb1764204/eras/babbage/impl/src/Cardano/Ledger/Babbage/Rules/Utxo.hs#L32

What about shelley? I know I'm stating the obvious, but 1.35.x is showing problems in shelley.

**Comment by JaredCorduan** [comment]
> What about shelley? I know I'm stating the obvious, but 1.35.x is showing problems in shelley.

The bug that I've found does not effect Shelley. What makes you think Shelley has problems? I'm not aware of a network still running Shelley...

**Comment by feqifei** [comment]
Great, downgrading to 1.34.1 and waiting for 1.35.3 :)

**Comment by gitmachtl** [comment]
Lets wait for 1.36.0 😆

**Comment by JaredCorduan** [comment]
thank you @reqlez, that is very kind of you, but please don't wait to downgrade on my/our account. downgrading to 1.34.1 on mainnet is indeed the best thing for everyone to do at the moment.

**Comment by reqlez** [comment]
> thank you @reqlez, that is very kind of you, but please don't wait to downgrade on my/our account. downgrading to 1.34.1 on mainnet is indeed the best thing for everyone to do at the moment.

I have a day job as well... sorry... i wish i could go 100% Cardano ;-)

**Comment by nemo83** [comment]
> > What about shelley? I know I'm stating the obvious, but 1.35.x is showing problems in shelley.
> 
> The bug that I've found does not effect Shelley. What makes you think Shelley has problems? I'm not aware of a network still running Shelley...

lool sorry, Alonzo! I meant Alonzo :disappear:

**Comment by karknu** [comment]
> > @papacarp I note that pooltool.io/realtime is giving surprisingly number of nodes for certain blocks (i.e less than ten) - this seems unusual.
> > Does anyone have evidence (i.e log entries) that show that any of these "height battles" above occurred _when_ the superseded block had been received by the about-to-forge node? My working hypothesis would be some sort of connectivity partition, I'm looking for evidence that assumption is incorrect.
> 
> I watched a block get 50 reports and then like 10 seconds later a new block came in and it orphaned the first block. Seems like the 1.34 nodes never got the 1.35 block so we only got reports from the 1.35 nodes. Then once a new block came in the 1.34 nodes moved the chain forward.
> 
> I think a network partition is the right angle to explore. 1.34 nodes stop talking to 1.35 nodes.

That should be easy to test by configuring a 1.34 relay to only have a single 1.35 node as its upstream peer.

**Comment by JaredCorduan** [comment]
I did write "resolves #2936 in the ledger PR notes that I think fixes the bug, but I did not realize that those magic words would close issues _cross-repo_. I think it's best to wait for more testing before this node issues is closed. sorry.

**Comment by JaredCorduan** [comment]
> So what we have is a soft fork where the "conservative nodes" (ie 1.34) are winning the longest chain.

I want to clarify that what I said here :point_up: is incorrect. This was actually an accidental _hard_ fork. Blocks produced by nodes with the bug would potentially _not_ be valid according to a 1.34 node. If the bug had instead been that a _higher_ fee was demanded, then it would have been a soft fork, since 1.34 nodes would still validate all the blocks from nodes with the bug.

**Comment by dorin100** [comment]
@gitmachtl is it ok to close this issue now?

**Comment by gitmachtl** [comment]
Yes, was resolved with 1.35.3. Thx.
